website/docs/reference/resource-configs/sql_header.md

```diff
@@ -2,25 +2,130 @@
---->
+<File name='models/<modelname>.sql'>
 
+```sql
+{{ config(
+  sql_header="<sql-statement>"
+) }}
+
+select ...
+
+
+```
+
+</File>
+
+<File name='dbt_project.yml'>
+
+```yml
+[config-version](config-version): 2
+
+models:
+  [<resource-path>](resource-path):
+    +sql_header: <sql-statement>
+
+```
+
+</File>
+
+</TabItem>
+
+
+<TabItem value="seeds">
+
+This config is not implemented for seeds
+
+</TabItem>
+
+<TabItem value="snapshots">
+
+<File name='snapshots/<filename>.sql'>
+
+```sql
+{% snapshot [snapshot_name](snapshot_name) %}
+
+{{ config(
+  sql_header="<sql-statement>"
+) }}
+
+select ...
+
+{% endsnapshot %}
+
+```
+
+</File>
+
+<File name='dbt_project.yml'>
+
+```yml
+snapshots:
+  [<resource-path>](resource-path):
+    +sql_header: <sql-statement>
+
+```
+
+</File>
+
+</TabItem>
+
+</Tabs>
+
+
+## Definition
+An optional configuration to inject SQL above the `create table as` and `create view as` statements that dbt executes when building models and snapshots.
 
-### Overview
+`sql_header`s can be set using the config, or by `call`-ing the `set_sql_header` macro (example below).
 
-dbt supports the specification of SQL headers: arbitrary SQL snippets that are injected above `create table as` and `create view as` statements that dbt executes.
+## Comparison to pre-hooks
+[Pre-hooks](pre-hook-post-hook) also provide an opportunity to execute SQL before model creation, as a _preceding_ query. In comparison, SQL in a `sql_header` is run in the same _query_ as the `create table|view as` statement.
 
-One possible use-case for SQL headers is the specification of [BigQuery Temporary UDFs](https://cloud.google.com/bigquery/docs/reference/standard-sql/user-defined-functions#sql-udf-examples). SQL headers are logically similar to model pre-hooks, with the notable difference being that they run in the same *query* as the `create table|view as` statement.
+As a result, this makes it more useful for [Snowflake session parameters](https://docs.snowflake.com/en/sql-reference/parameters.html) and [BigQuery Temporary UDFs](https://cloud.google.com/bigquery/docs/reference/standard-sql/user-defined-functions#sql-udf-examples).
 
-### Usage
+## Examples
+
+### Set Snowflake session parameters for a particular model
+This uses the config block syntax:
+<File name='models/my_model.sql'>
+
+```sql
+{{ config(
+  sql_header="alter session set timezone = 'Australia/Sydney'"
+) }}
+
+select * from {{ ref('other_model') }}
+```
+
+</File>
+
+### Set Snowflake session parameters for all models
+
+<File name='dbt_project.yml'>
+
+```yml
+config-version: 2
+
+models:
+  +sql_header: "alter session set timezone = 'Australia/Sydney'"
+```
+
+</File>
+
+### Create a BigQuery Temporary UDF
+
+This example calls the `set_sql_header` macro. This macro is a convenience wrapper which you may choose to use if you have a multi-line SQL statement to inject. You do not need to use the `sql_header` configuration key in this case.
```

Comment:
it's still not clear to me why we provide this method. Why not just do:
```
{% set my_bq_udf %}
-- this is sql
{% endset %}

{{ config(sql_header = my_bq_udf) }}

```

Reply:
No real reason, other than that's how the original contributor wanted to use it (https://github.com/fishtown-analytics/dbt/pull/1967)